File: top100/94_binary_tree_inorder/binaryTreeInorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    /* Inorder: left-root-right */
    /* stack maintains the execution order in recursion */
    public List<Integer> inorderTraversal(TreeNode root) 
    {
        List<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        
        TreeNode curr = root;
        while (curr != null)
        {
            if (curr.left != null)  // add left branch all the way down
            {
                stack.push(curr);
                curr = curr.left;
            }
            else                    // reaches the left leaf
            {
                result.add(curr.val);   // add left leaf to result
                while (!stack.isEmpty() && curr.right == null)  
                    // backtrack to root until a right presents
                {
                    TreeNode currParent = stack.pop();
                    result.add(currParent.val); // indorder add root in middle
                    curr = currParent;
                }
                curr = curr.right;              // visit right last
            }
        }        
        return result;
    }
    
}

File: top100/94_binary_tree_inorder/binaryTreeInorder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        stack = []
        
        curr = root
        while curr is not None:
            if curr.left is not None:
                stack.append(curr)
                curr = curr.left
            else:
                result.append(curr.val)
                while len(stack) != 0 and curr.right is None:
                    node = stack.pop()
                    result.append(node.val)
                    curr = node
                curr = curr.right
        
        return result